crates/web/src/lib.rs

```diff
@@ -59,10 +59,216 @@ impl<'a> rose::FuncNode for &'a Func {
+                }
+                rose::Expr::Accum { var, vector, body } => {
+                    writeln!(&mut s, "accum x{}: T{} {{", var.var(), vector.ty())?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    let x = def.blocks[body.block()].arg.var();
+                    writeln!(&mut s, "  x{x}: T{}", def.vars[x].ty())?;
+                    print_block(s, def, spaces + 2, *body)?;
+                    for _ in 0..spaces {
+                        write!(&mut s, " ")?;
+                    }
+                    writeln!(&mut s, "}}")?
+                }
+                rose::Expr::Add { accum, addend } => {
+                    writeln!(&mut s, "x{} += x{}", accum.var(), addend.var())?
+                }
+            }
+        }
+        for _ in 0..spaces {
+            write!(&mut s, " ")?;
+        }
+        writeln!(&mut s, "x{}", def.blocks[b.block()].ret.var())?;
+        Ok(())
+    }
+
+    fn print_elems(s: &mut String, items: &Vec<rose::id::Var>) -> std::fmt::Result {
+        let mut first = true;
+        for item in items {
+            if first {
+                first = false;
+            } else {
+                write!(s, ", ")?;
+            }
+            write!(s, "x{}", item.var())?;
+        }
+        Ok(())
+    }
+
+    let mut s = String::new();
     let (_, def) = f.rc.as_ref();
-    format!("{:#?}", def)
+
+    for (i, constraints) in def.generics.iter().enumerate() {
+        write!(&mut s, "G{i} = ")?;
+        let mut first = true;
+        for constraint in constraints.iter() {
+            if first {
+                first = false;
+            } else {
+                write!(&mut s, " + ")?;
+            }
+            write!(&mut s, "{constraint:?}")?;
+        }
+        writeln!(&mut s)?;
+    }
+    for (i, ty) in def.types.iter().enumerate() {
+        write!(&mut s, "T{i} = ")?;
+        match ty {
+            rose::Ty::Unit | rose::Ty::Bool | rose::Ty::F64 => writeln!(&mut s, "{ty:?}")?,
+            rose::Ty::Fin { size } => writeln!(&mut s, "{size}")?,
+            rose::Ty::Generic { id } => writeln!(&mut s, "G{}", id.generic())?,
+            rose::Ty::Scope { id } => writeln!(&mut s, "B{}", id.block())?,
+            rose::Ty::Ref { scope, inner } => {
+                writeln!(&mut s, "Ref T{} T{}", scope.ty(), inner.ty())?
+            }
+            rose::Ty::Array { index, elem } => {
+                writeln!(&mut s, "[T{}; T{}]", elem.ty(), index.ty())?
+            }
+            rose::Ty::Tuple { members } => {
+                write!(&mut s, "(")?;
+                let mut first = true;
+                for member in members {
+                    if first {
+                        first = false;
+                    } else {
+                        write!(&mut s, ", ")?;
+                    }
+                    write!(&mut s, "T{}", member.ty())?;
+                }
```

Comment:
Could you make the `print_elems` function a bit more general and use it to replace this chunk of code? Maybe take in a prefix character and add an [`impl`](https://doc.rust-lang.org/std/keyword.impl.html) parameter for an [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html) whose `Item` is [`usize`](https://doc.rust-lang.org/std/primitive.usize.html), and in this usage, [`map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map) over `members` to turn each one into a `usize`.